File: packages/client/ios/RNOta/Store/BundleStore.swift
import Foundation

/// M2 — Bundle Store (iOS).
///
/// Owns on-disk OTA layout and atomic `state.json` transitions.
/// Deliberately unaware of networking, crypto, and React Native bundle loading.
///
/// Thread-safe via an internal lock. Feature-equivalent to Android `BundleStore`.
public final class BundleStore: @unchecked Sendable {
  private let config: BundleStoreConfig
  private let io: FileIo
  private let lock = NSRecursiveLock()

  private let otaDir: URL
  private let stateFile: URL
  private let stateTempFile: URL
  private let bootFile: URL
  private let slotsDir: URL

  private var cachedState: BundleState = .empty

  public init(config: BundleStoreConfig) {
    self.config = config
    self.io = config.fileIo
    self.otaDir = config.otaDirectory
    self.stateFile = otaDir.appendingPathComponent(OtaPaths.stateFileName)
    self.stateTempFile = otaDir.appendingPathComponent(OtaPaths.stateTempFileName)
    self.bootFile = otaDir.appendingPathComponent(OtaPaths.bootFileName)
    self.slotsDir = otaDir.appendingPathComponent(OtaPaths.slotsDirName)
  }

  public func initialize() throws {
    try withLock {
      try ensureLayout()
      try discardIncompleteTemp()
      cachedState = try readAndRecoverLocked()
    }
  }

  /// Reads only `state.json` (plus recovery rewrite if needed). Does not scan `slots/`.
  @discardableResult
  public func loadState() throws -> BundleState {
    try withLock {
      try ensureLayout()
      try discardIncompleteTemp()
      cachedState = try readAndRecoverLocked()
      return cachedState
    }
  }

  public func saveState(_ state: BundleState) throws {
    try withLock {
      try ensureLayout()
      let normalized = BundleState(
        schemaVersion: config.supportedSchemaVersion,
        installedBinaryVersion: state.installedBinaryVersion ?? config.installedBinaryVersion,
        activeSlot: OtaStateCodec.sanitizeSlotId(state.activeSlot),
        pendingSlot: OtaStateCodec.sanitizeSlotId(state.pendingSlot),
        previousSlot: OtaStateCodec.sanitizeSlotId(state.previousSlot)
      )
      try writeStateLocked(normalized)
      cachedState = normalized
    }
  }

  /// Creates slot dirs + `assets/`. Never writes `bundle.hbc` / never touches `state.json`.
  @discardableResult
  public func createSlot(_ slotId: String) throws -> SlotInfo {
    try withLock {
      try requireValidSlotId(slotId)
      let dir = slotDirectory(slotId)
      if io.fileExists(dir) {
        throw BundleStoreException.slotAlreadyExists(slotId)
      }
      let assets = dir.appendingPathComponent(OtaPaths.assetsDirName)
      do {
        try io.createDirectory(dir)
        try io.createDirectory(assets)
      } catch {
        throw BundleStoreException.ioFailure("Failed to create slot directories for \(slotId)")
      }
      return slotInfoLocked(slotId)
    }
  }

  public func deleteSlot(_ slotId: String) throws {
    try withLock {
      try requireValidSlotId(slotId)
      if cachedState.referencedSlotIds().contains(slotId) {
        throw BundleStoreException.slotInUse(slotId)
      }
      let dir = slotDirectory(slotId)
      guard io.fileExists(dir) else {
        throw BundleStoreException.slotNotFound(slotId)
      }
      do {
        try io.removeItem(dir)
      } catch {
        throw BundleStoreException.ioFailure("Failed to delete slot \(slotId)")
      }
    }
  }

  public func slotExists(_ slotId: String) -> Bool {
    withLock {
      guard OtaStateCodec.isValidSlotId(slotId) else { return false }
      return io.isDirectory(slotDirectory(slotId))
    }
  }

  /// Explicit directory scan — call only when listing is required.
  public func listSlots() throws -> [SlotInfo] {
    try withLock {
      try ensureLayout()
      let names = try io.listDirectoryNames(slotsDir)
      return names
        .filter { OtaStateCodec.isValidSlotId($0) }
        .sorted()
        .map { slotInfoLocked($0) }
    }
  }

  public func setActiveSlot(_ slotId: String?) throws {
    try withLock {
      let next: BundleState
      if let slotId {
        try requireValidSlotId(slotId)
        try requireCommittedSlot(slotId)
        next = withBinaryVersion(cachedState.withActive(slotId))
      } else {
        next = withBinaryVersion(cachedState.withActive(nil))
      }
      try writeStateLocked(next)
      cachedState = next
    }
  }

  public func setPendingSlot(_ slotId: String?) throws {
    try withLock {
      let next: BundleState
      if let slotId {
        try requireValidSlotId(slotId)
        try requireCommittedSlot(slotId)
        next = withBinaryVersion(cachedState.withPending(slotId))
      } else {
        next = withBinaryVersion(cachedState.withPending(nil))
      }
      try writeStateLocked(next)
      cachedState = next
    }
  }

  public func clearPendingSlot() throws {
    try setPendingSlot(nil)
  }

  public func setPreviousSlot(_ slotId: String?) throws {
    try withLock {
      let next: BundleState
      if let slotId {
        try requireValidSlotId(slotId)
        try requireCommittedSlot(slotId)
        next = withBinaryVersion(cachedState.withPrevious(slotId))
      } else {
        next = withBinaryVersion(cachedState.withPrevious(nil))
      }
      try writeStateLocked(next)
      cachedState = next
    }
  }

  @discardableResult
  public func garbageCollect() throws -> [String] {
    try withLock {
      try ensureLayout()
      let keep = cachedState.referencedSlotIds()
      var deleted: [String] = []
      for name in try io.listDirectoryNames(slotsDir) {
        if !OtaStateCodec.isValidSlotId(name) {
          try? io.removeItem(slotsDir.appendingPathComponent(name))
          deleted.append(name)
          continue
        }
        if !keep.contains(name) {
          try? io.removeItem(slotsDir.appendingPathComponent(name))
          deleted.append(name)
        }
      }
      return deleted.sorted()
    }
  }

  public func slotDirectory(_ slotId: String) -> URL {
    slotsDir.appendingPathComponent(slotId)
  }

  public func bundlePath(_ slotId: String) -> URL {
    slotDirectory(slotId).appendingPathComponent(OtaPaths.bundleFileName)
  }

  public func manifestPath(_ slotId: String) -> URL {
    slotDirectory(slotId).appendingPathComponent(OtaPaths.manifestFileName)
  }

  public func assetsDirectory(_ slotId: String) -> URL {
    slotDirectory(slotId).appendingPathComponent(OtaPaths.assetsDirName)
  }

  public func currentState() -> BundleState {
    withLock { cachedState }
  }

  /// Read-only snapshot of `state.json` for M1 Bundle Resolver.
  /// Does **not** create layout, recover, or write.
  public func peekState() -> BundleState {
    withLock {
      guard io.isFile(stateFile) else { return .empty }
      do {
        let bytes = try io.readData(from: stateFile)
        return OtaStateCodec.decode(bytes: bytes, supportedSchemaVersion: config.supportedSchemaVersion)
          ?? .empty
      } catch {
        return .empty
      }
    }
  }

  /// True when `slots/<id>/bundle.hbc` exists as a regular file. Read-only.
  public func hasCommittedBundle(_ slotId: String) -> Bool {
    withLock {
      OtaStateCodec.isValidSlotId(slotId) && isCommittedLocked(slotId)
    }
  }

  // MARK: - Internals

  private func ensureLayout() throws {
    do {
      try io.createDirectory(otaDir)
      try io.createDirectory(slotsDir)
    } catch {
      throw BundleStoreException.ioFailure("Cannot create OTA directories")
    }
    if !io.fileExists(bootFile) {
      let bootTemp = otaDir.appendingPathComponent("\(OtaPaths.bootFileName).tmp")
      try io.writeAtomic(target: bootFile, temporary: bootTemp, bytes: Data("{}\n".utf8))
    }
  }

  private func discardIncompleteTemp() throws {
    if io.fileExists(stateTempFile) {
      try? io.removeItem(stateTempFile)
    }
  }

  private func readAndRecoverLocked() throws -> BundleState {
    let loaded: BundleState?
    if io.isFile(stateFile) {
      loaded = try? OtaStateCodec.decode(
        bytes: io.readData(from: stateFile),
        supportedSchemaVersion: config.supportedSchemaVersion
      )
    } else {
      loaded = nil
    }

    let base = loaded ?? BundleState(
      schemaVersion: config.supportedSchemaVersion,
      installedBinaryVersion: config.installedBinaryVersion
    )
    let recovered = sanitizeReferences(base)
    let needsRewrite =
      loaded == nil ||
      recovered != base ||
      recovered.schemaVersion != config.supportedSchemaVersion

    let toPersist = BundleState(
      schemaVersion: config.supportedSchemaVersion,
      installedBinaryVersion: recovered.installedBinaryVersion ?? config.installedBinaryVersion,
      activeSlot: recovered.activeSlot,
      pendingSlot: recovered.pendingSlot,
      previousSlot: recovered.previousSlot
    )

    if needsRewrite {
      try writeStateLocked(toPersist)
    }
    return toPersist
  }

  private func sanitizeReferences(_ state: BundleState) -> BundleState {
    func resolve(_ id: String?) -> String? {
      guard let id else { return nil }
      guard let sanitized = OtaStateCodec.sanitizeSlotId(id) else { return nil }
      return isCommittedLocked(sanitized) ? sanitized : nil
    }
    return BundleState(
      schemaVersion: state.schemaVersion,
      installedBinaryVersion: state.installedBinaryVersion,
      activeSlot: resolve(state.activeSlot),
      pendingSlot: resolve(state.pendingSlot),
      previousSlot: resolve(state.previousSlot)
    )
  }

  private func writeStateLocked(_ state: BundleState) throws {
    let bytes = OtaStateCodec.encode(state)
    try io.writeAtomic(target: stateFile, temporary: stateTempFile, bytes: bytes)
  }

  private func withBinaryVersion(_ state: BundleState) -> BundleState {
    BundleState(
      schemaVersion: config.supportedSchemaVersion,
      installedBinaryVersion: state.installedBinaryVersion ?? config.installedBinaryVersion,
      activeSlot: state.activeSlot,
      pendingSlot: state.pendingSlot,
      previousSlot: state.previousSlot
    )
  }

  private func requireValidSlotId(_ slotId: String) throws {
    guard OtaStateCodec.isValidSlotId(slotId) else {
      throw BundleStoreException.invalidSlotId(slotId)
    }
  }

  private func requireCommittedSlot(_ slotId: String) throws {
    guard io.isDirectory(slotDirectory(slotId)) else {
      throw BundleStoreException.slotNotFound(slotId)
    }
    guard isCommittedLocked(slotId) else {
      throw BundleStoreException.slotNotCommitted(slotId)
    }
  }

  private func isCommittedLocked(_ slotId: String) -> Bool {
    io.isFile(bundlePath(slotId))
  }

  private func slotInfoLocked(_ slotId: String) -> SlotInfo {
    SlotInfo(
      id: slotId,
      directory: slotDirectory(slotId),
      bundleFile: bundlePath(slotId),
      manifestFile: manifestPath(slotId),
      assetsDirectory: assetsDirectory(slotId),
      isCommitted: isCommittedLocked(slotId)
    )
  }

  private func withLock<T>(_ body: () throws -> T) rethrows -> T {
    lock.lock()
    defer { lock.unlock() }
    return try body()
  }
}

private extension BundleState {
  func withActive(_ id: String?) -> BundleState {
    BundleState(
      schemaVersion: schemaVersion,
      installedBinaryVersion: installedBinaryVersion,
      activeSlot: id,
      pendingSlot: pendingSlot,
      previousSlot: previousSlot
    )
  }

  func withPending(_ id: String?) -> BundleState {
    BundleState(
      schemaVersion: schemaVersion,
      installedBinaryVersion: installedBinaryVersion,
      activeSlot: activeSlot,
      pendingSlot: id,
      previousSlot: previousSlot
    )
  }

  func withPrevious(_ id: String?) -> BundleState {
    BundleState(
      schemaVersion: schemaVersion,
      installedBinaryVersion: installedBinaryVersion,
      activeSlot: activeSlot,
      pendingSlot: pendingSlot,
      previousSlot: id
    )
  }
}

File: packages/client/android/src/main/java/com/rnota/store/BundleStore.kt
package com.rnota.store

import java.io.File
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * M2 — Bundle Store.
 *
 * Owns on-disk OTA layout and atomic `state.json` transitions.
 * Deliberately unaware of networking, crypto, and React Native bundle loading.
 *
 * Thread-safe: all public mutations take an internal lock.
 */
class BundleStore(
  private val config: BundleStoreConfig,
) {
  private val lock = ReentrantLock()
  private val io: FileIo = config.fileIo

  private val otaDir: File = config.otaDirectory
  private val stateFile: File = File(otaDir, OtaPaths.STATE_FILE_NAME)
  private val stateTempFile: File = File(otaDir, OtaPaths.STATE_TEMP_FILE_NAME)
  private val bootFile: File = File(otaDir, OtaPaths.BOOT_FILE_NAME)
  private val slotsDir: File = File(otaDir, OtaPaths.SLOTS_DIR_NAME)

  @Volatile
  private var cachedState: BundleState = BundleState.EMPTY

  /** Creates `<ota>/`, `slots/`, `boot.json`, and recovers `state.json` into a valid snapshot. */
  fun initialize() =
    lock.withLock {
      ensureLayout()
      discardIncompleteTemp()
      cachedState = readAndRecoverLocked()
    }

  /** Reads only `state.json` (plus recovery rewrite if needed). Does not scan `slots/`. */
  fun loadState(): BundleState =
    lock.withLock {
      ensureLayout()
      discardIncompleteTemp()
      cachedState = readAndRecoverLocked()
      cachedState
    }

  /** Atomically persists [state] (write temp → fsync → rename). */
  fun saveState(state: BundleState) =
    lock.withLock {
      ensureLayout()
      val normalized =
        state.copy(
          schemaVersion = config.supportedSchemaVersion,
          installedBinaryVersion = state.installedBinaryVersion ?: config.installedBinaryVersion,
          activeSlot = OtaStateCodec.sanitizeSlotId(state.activeSlot),
          pendingSlot = OtaStateCodec.sanitizeSlotId(state.pendingSlot),
          previousSlot = OtaStateCodec.sanitizeSlotId(state.previousSlot),
        )
      writeStateLocked(normalized)
      cachedState = normalized
    }

  /**
   * Creates `slots/<id>/` with `assets/`. Does **not** create `bundle.hbc` / `manifest.json`
   * (those are written by Downloader / Verifier). Never touches `state.json`.
   */
  fun createSlot(slotId: String): SlotInfo =
    lock.withLock {
      requireValidSlotId(slotId)
      val dir = slotDirectory(slotId)
      if (io.exists(dir)) {
        throw BundleStoreException.SlotAlreadyExists(slotId)
      }
      val assets = File(dir, OtaPaths.ASSETS_DIR_NAME)
      if (!io.mkdir(dir) || !io.mkdir(assets)) {
        throw BundleStoreException.IoFailure("Failed to create slot directories for $slotId")
      }
      slotInfoLocked(slotId)
    }

  fun deleteSlot(slotId: String) =
    lock.withLock {
      requireValidSlotId(slotId)
      val state = cachedState
      if (slotId in state.referencedSlotIds()) {
        throw BundleStoreException.SlotInUse(slotId)
      }
      val dir = slotDirectory(slotId)
      if (!io.exists(dir)) {
        throw BundleStoreException.SlotNotFound(slotId)
      }
      if (!io.deleteRecursively(dir)) {
        throw BundleStoreException.IoFailure("Failed to delete slot $slotId")
      }
    }

  fun slotExists(slotId: String): Boolean =
    lock.withLock {
      if (!OtaStateCodec.isValidSlotId(slotId)) {
        return false
      }
      io.isDirectory(slotDirectory(slotId))
    }

  /** Explicit directory scan — call only when listing is required. */
  fun listSlots(): List<SlotInfo> =
    lock.withLock {
      ensureLayout()
      io.listNames(slotsDir)
        .filter { OtaStateCodec.isValidSlotId(it) }
        .sorted()
        .map { slotInfoLocked(it) }
    }

  /**
   * Explicitly marks [slotId] as active. Requires a **committed** slot (`bundle.hbc` present).
   * Pass null to clear (embedded fallback). Does not auto-promote pending.
   */
  fun setActiveSlot(slotId: String?) =
    lock.withLock {
      val next =
        when (slotId) {
          null -> cachedState.copy(activeSlot = null)
          else -> {
            requireValidSlotId(slotId)
            requireCommittedSlot(slotId)
            cachedState.copy(activeSlot = slotId)
          }
        }
      writeStateLocked(withBinaryVersion(next))
      cachedState = withBinaryVersion(next)
    }

  fun setPendingSlot(slotId: String?) =
    lock.withLock {
      val next =
        when (slotId) {
          null -> cachedState.copy(pendingSlot = null)
          else -> {
            requireValidSlotId(slotId)
            requireCommittedSlot(slotId)
            cachedState.copy(pendingSlot = slotId)
          }
        }
      writeStateLocked(withBinaryVersion(next))
      cachedState = withBinaryVersion(next)
    }

  fun clearPendingSlot() {
    setPendingSlot(null)
  }

  fun setPreviousSlot(slotId: String?) =
    lock.withLock {
      val next =
        when (slotId) {
          null -> cachedState.copy(previousSlot = null)
          else -> {
            requireValidSlotId(slotId)
            requireCommittedSlot(slotId)
            cachedState.copy(previousSlot = slotId)
          }
        }
      writeStateLocked(withBinaryVersion(next))
      cachedState = withBinaryVersion(next)
    }

  /**
   * Deletes slot directories not referenced by active / pending / previous.
   * Retention policy: active + previous + at most one pending (ARCHITECTURE.md M2).
   */
  fun garbageCollect(): List<String> =
    lock.withLock {
      ensureLayout()
      val keep = cachedState.referencedSlotIds()
      val deleted = mutableListOf<String>()
      for (name in io.listNames(slotsDir)) {
        if (!OtaStateCodec.isValidSlotId(name)) {
          // Non-conforming directory names are removed defensively.
          io.deleteRecursively(File(slotsDir, name))
          deleted.add(name)
          continue
        }
        if (name !in keep) {
          io.deleteRecursively(File(slotsDir, name))
          deleted.add(name)
        }
      }
      deleted.sorted()
    }

  /** Path helpers for other modules (Downloader writes into these). */
  fun slotDirectory(slotId: String): File = File(slotsDir, slotId)

  fun bundlePath(slotId: String): File = File(slotDirectory(slotId), OtaPaths.BUNDLE_FILE_NAME)

  fun manifestPath(slotId: String): File = File(slotDirectory(slotId), OtaPaths.MANIFEST_FILE_NAME)

  fun assetsDirectory(slotId: String): File = File(slotDirectory(slotId), OtaPaths.ASSETS_DIR_NAME)

  fun currentState(): BundleState = lock.withLock { cachedState }

  // region internals

  private fun ensureLayout() {
    if (!io.mkdir(otaDir)) {
      throw BundleStoreException.IoFailure("Cannot create OTA directory: $otaDir")
    }
    if (!io.mkdir(slotsDir)) {
      throw BundleStoreException.IoFailure("Cannot create slots directory: $slotsDir")
    }
    if (!io.exists(bootFile)) {
      // M6 owns boot.json semantics; we only ensure the path exists.
      try {
        io.writeAtomic(bootFile, File(otaDir, "${OtaPaths.BOOT_FILE_NAME}.tmp"), "{}\n".toByteArray())
      } catch (e: BundleStoreException) {
        throw e
      } catch (e: Exception) {
        throw BundleStoreException.IoFailure("Cannot create boot.json", e)
      }
    }
  }

  private fun discardIncompleteTemp() {
    if (io.exists(stateTempFile)) {
      io.deleteRecursively(stateTempFile)
    }
  }

  private fun readAndRecoverLocked(): BundleState {
    val loaded =
      if (!io.exists(stateFile) || !io.isFile(stateFile)) {
        null
      } else {
        try {
          OtaStateCodec.decode(io.readBytes(stateFile), config.supportedSchemaVersion)
        } catch (e: Exception) {
          null
        }
      }

    val base = loaded ?: BundleState.EMPTY.copy(installedBinaryVersion = config.installedBinaryVersion)
    val recovered = sanitizeReferences(base)
    val needsRewrite =
      loaded == null ||
        recovered != base ||
        recovered.schemaVersion != config.supportedSchemaVersion

    val toPersist =
      recovered.copy(
        schemaVersion = config.supportedSchemaVersion,
        installedBinaryVersion = recovered.installedBinaryVersion ?: config.installedBinaryVersion,
      )

    if (needsRewrite) {
      writeStateLocked(toPersist)
    }
    return toPersist
  }

  private fun sanitizeReferences(state: BundleState): BundleState {
    fun resolve(id: String?): String? {
      if (id == null) return null
      val sanitized = OtaStateCodec.sanitizeSlotId(id) ?: return null
      return if (isCommittedLocked(sanitized)) sanitized else null
    }
    return state.copy(
      activeSlot = resolve(state.activeSlot),
      pendingSlot = resolve(state.pendingSlot),
      previousSlot = resolve(state.previousSlot),
    )
  }

  private fun writeStateLocked(state: BundleState) {
    val bytes = OtaStateCodec.encode(state)
    io.writeAtomic(stateFile, stateTempFile, bytes)
  }

  private fun withBinaryVersion(state: BundleState): BundleState =
    state.copy(
      schemaVersion = config.supportedSchemaVersion,
      installedBinaryVersion = state.installedBinaryVersion ?: config.installedBinaryVersion,
    )

  private fun requireValidSlotId(slotId: String) {
    if (!OtaStateCodec.isValidSlotId(slotId)) {
      throw BundleStoreException.InvalidSlotId(slotId)
    }
  }

  private fun requireCommittedSlot(slotId: String) {
    if (!io.isDirectory(slotDirectory(slotId))) {
      throw BundleStoreException.SlotNotFound(slotId)
    }
    if (!isCommittedLocked(slotId)) {
      throw BundleStoreException.SlotNotCommitted(slotId)
    }
  }

  private fun isCommittedLocked(slotId: String): Boolean = io.isFile(bundlePath(slotId))

  private fun slotInfoLocked(slotId: String): SlotInfo {
    val dir = slotDirectory(slotId)
    return SlotInfo(
      id = slotId,
      directory = dir,
      bundleFile = bundlePath(slotId),
      manifestFile = manifestPath(slotId),
      assetsDirectory = assetsDirectory(slotId),
      isCommitted = isCommittedLocked(slotId),
    )
  }

  // endregion
}
